### DISCUSSION TITLE
Pixel grid for pixel art

### DISCUSSION THREAD

**Discussion by oliver-dreamwalker** [opentoonz/opentoonz#6351] [Ideas]
### Is your feature request related to a problem? Please describe

I think there are people that use Opentoonz for Gamedev (animated sprites, pixel art)
and they would benefit by the addition of a pixel grid.

### Describe the solution you'd like

A pixel grid that disappear when zoomed out
and it reappears when zoomed in

![pixel](https://github.com/opentoonz/opentoonz/assets/66680990/ef9b1442-42dc-49bc-9a46-f067cc268cd6)
![pixel2 ](https://github.com/opentoonz/opentoonz/assets/66680990/2e725dc1-68b9-42dc-846c-62b8ab1c7d6f)

**Comment by oliver-dreamwalker** [comment]
Yeah, Asprite is superior at pixel art compared to OpenToonz, the same way ZBrush is superior at 3d sculpting compared to Blender.

But a pixel grid would still be a nice feature.

**Comment by oliver-dreamwalker** [comment]
> OT has virtually nothing on that front. The GIMP has more pixeling capacity than OT.

OpenToonz has the "pencil"(pixel) mode and it is capable of exporting a sprite-sheet, so I do not think it is **that bad**, 
but I think **Pixelorama** technically is better for pixel art, although, I think OT's timeline better.

I have seen some people use OT for pixel animation, so it's possible.


An adjustable grid is also a good idea.

**Comment by andeon** [comment]
@oliver-dreamwalker I'd like to see a feature like this first in Tahoma2d. @manongjohn  created a nice grid and overlay feature, so adding a "pixel grid" would be a great addition. However, it should function similar to the "Field Guide," serving as a guide without interacting with the brush, correct?

**Comment by oliver-dreamwalker** [comment]
>serving as a guide without interacting with the brush, correct?

Yeah, I guess so.

**Comment by RodneyBaker** [comment]
Not suggesting that this feature shouldn't be added.  It would be quite welcome.
This morning I found myself playing with a workaround that uses images as a checkerboard... and then I copied your grid and added that over the top of that checkerboard.  It worked pretty well.

The thought being that we have several things in Opentoonz right now that are workarounds and near hits for the feature.
A near hit for instance example is the grid (chessboard) that is displayed for transparency.
Unfortunately there are two shortfalls with that currently; 

- The transparency chessboard can only be used with preview on (as far as I can tell
- The squares are 50px by 50px (I think)

For pixel art I don't think it would be difficult to draw with Preview on.
An option to allow changing of chessboard square size would have to be implemented.

The other option I tested that looked promising (again as a workaround) is using textures themselves to fill a separate level/column and then use that under the level being drawn on.

At any rate, it was a fun exercise.

![image](https://github.com/opentoonz/opentoonz/assets/2576432/4c943ede-de5f-4d5d-93a9-1ac4808fd07a)

![image](https://github.com/opentoonz/opentoonz/assets/2576432/39c2848b-2bc6-447c-9f1e-c89652eff150)



Added:  I was finding that erasing at pixel art level doesn't work very well.  Drawing with Color Style 0 seemed to work better for erasing.

![pixelcheckboardandgridthemovie](https://github.com/opentoonz/opentoonz/assets/2576432/54cfd8f6-120c-4b2b-9412-46f8e39a8670)

**Comment by andeon** [comment]
I see this possible create a grid in two ways.

1- Using 'Safe Area' is required to edit safearea.ini at /stuff/config  and insert a new predefined group parameters, it doesn't work well with different resolutions. So I'm avoiding focus on this option for now. 

But something looks like this:
```
[SafeArea5]
area\0=100, 0
area\1=100, 0.185, 0, 100, 255
area\2=100, 0.370, 0, 100, 255
...
area\64=3.230, 100, 0, 100, 255
area\65=3.329, 100
name=Pixel_grid
```

2- Using 'Guides' on 'Viewer'.

Editing a file .tnz for e.g. and add it bellow `<properties>`

**64x64 e.g.**
```
    <hGuides>
      -14.222208 -13.777764 -13.333320 -12.888876 -12.444432 -11.999988 -11.555544 -11.111100 -10.666656 -10.222212 -9.777768 -9.333324 -8.888880 -8.444436 -7.999992 -7.555548 -7.111104 -6.666660 -6.222216 -5.777772 -5.333328 -4.888884 -4.444440 -3.555552 -3.999996 -3.111108 -2.666664 -2.222220 -1.777776 -1.33332 -0.888888 -0.444444 0 0.444444 0.888888 1.33332 1.777776 2.222220 2.666664 3.111108 3.555552 3.999996 4.444440 4.888884 5.333328 5.777772 6.222216 6.666660 7.111104 7.555548 7.999992 8.444436 8.888880 9.333324 9.777768 10.222212 10.666656 11.111100 11.555544 11.999988 12.444432 12.888876 13.333320 13.777764 14.222208
    </hGuides>
    <vGuides>
      -14.222208 -13.777764 -13.333320 -12.888876 -12.444432 -11.999988 -11.555544 -11.111100 -10.666656 -10.222212 -9.777768 -9.333324 -8.888880 -8.444436 -7.999992 -7.555548 -7.111104 -6.666660 -6.222216 -5.777772 -5.333328 -4.888884 -4.444440 -3.555552 -3.999996 -3.111108 -2.666664 -2.222220 -1.777776 -1.33332 -0.888888 -0.444444 0 0.444444 0.888888 1.33332 1.777776 2.222220 2.666664 3.111108 3.555552 3.999996 4.444440 4.888884 5.333328 5.777772 6.222216 6.666660 7.111104 7.555548 7.999992 8.444436 8.888880 9.333324 9.777768 10.222212 10.666656 11.111100 11.555544 11.999988 12.444432 12.888876 13.333320 13.777764 14.222208
    </vGuides>
```
**128x128 e.g.**
```
    <hGuides>
      -28.444416 -27.999972 -27.555528 -27.111084 -26.66664 -26.222196 -25.77775 -25.333308 -24.888864 -24.44442 -23.999976 -23.555532 -23.111088 -22.666644 -22.2222 -21.777756 -21.333312 -20.888868 -20.444424 -19.99998 -19.555536 -19.111092 -18.666648 -18.222204 -17.77776 -17.333316 -16.888872 -16.444428 -15.999984 -15.55554 -15.111096 -14.666652 -14.222208 -13.777764 -13.333320 -12.888876 -12.444432 -11.999988 -11.555544 -11.111100 -10.666656 -10.222212 -9.777768 -9.333324 -8.888880 -8.444436 -7.999992 -7.555548 -7.111104 -6.666660 -6.222216 -5.777772 -5.333328 -4.888884 -4.444440 -3.555552 -3.999996 -3.111108 -2.666664 -2.222220 -1.777776 -1.33332 -0.888888 -0.444444 0 0.444444 0.888888 1.33332 1.777776 2.222220 2.666664 3.111108 3.555552 3.999996 4.444440 4.888884 5.333328 5.777772 6.222216 6.666660 7.111104 7.555548 7.999992 8.444436 8.888880 9.333324 9.777768 10.222212 10.666656 11.111100 11.555544 11.999988 12.444432 12.888876 13.333320 13.777764 14.222208 14.666652 15.111096 15.55554 15.999984 16.444428 16.888872 17.333316 17.77776 18.222204 18.666648 19.111092 19.555536 19.99998 20.444424 20.888868 21.333312 21.777756 22.2222 22.666644 23.111088 23.555532 23.999976 24.44442 24.888864 25.333308 25.777752 26.222196 26.66664 27.111084 27.555528 27.999972 28.444416
    </hGuides>
    <vGuides>
      -28.444416 -27.999972 -27.555528 -27.111084 -26.66664 -26.222196 -25.77775 -25.333308 -24.888864 -24.44442 -23.999976 -23.555532 -23.111088 -22.666644 -22.2222 -21.777756 -21.333312 -20.888868 -20.444424 -19.99998 -19.555536 -19.111092 -18.666648 -18.222204 -17.77776 -17.333316 -16.888872 -16.444428 -15.999984 -15.55554 -15.111096 -14.666652 -14.222208 -13.777764 -13.333320 -12.888876 -12.444432 -11.999988 -11.555544 -11.111100 -10.666656 -10.222212 -9.777768 -9.333324 -8.888880 -8.444436 -7.999992 -7.555548 -7.111104 -6.666660 -6.222216 -5.777772 -5.333328 -4.888884 -4.444440 -3.555552 -3.999996 -3.111108 -2.666664 -2.222220 -1.777776 -1.33332 -0.888888 -0.444444 0 0.444444 0.888888 1.33332 1.777776 2.222220 2.666664 3.111108 3.555552 3.999996 4.444440 4.888884 5.333328 5.777772 6.222216 6.666660 7.111104 7.555548 7.999992 8.444436 8.888880 9.333324 9.777768 10.222212 10.666656 11.111100 11.555544 11.999988 12.444432 12.888876 13.333320 13.777764 14.222208 14.666652 15.111096 15.55554 15.999984 16.444428 16.888872 17.333316 17.77776 18.222204 18.666648 19.111092 19.555536 19.99998 20.444424 20.888868 21.333312 21.777756 22.2222 22.666644 23.111088 23.555532 23.999976 24.44442 24.888864 25.333308 25.777752 26.222196 26.66664 27.111084 27.555528 27.999972 28.444416
    </vGuides>
```

It's a 64x64 pixel grid I created to show.

![grid](https://github.com/opentoonz/opentoonz/assets/11843239/0e131864-6909-477a-8464-13d96f402675)


Since I usually only work at 1920x1080, I saw that it's possible to use the Sub-camera to map a specific area to render in preview.

First I set the Sub-camera to 64x64 in the 'Viewer' (the guides zone). So, set View Settings > Use Sub-camera.

'Preview' lets you save the images like in render. If you have doubts about the extensions, changing them in the 'output settings', the 'Preview' follows the configuration. So could generate a 'Spritesheet' for example.

The Sub-camera can be very flexible and I had forgotten about that.

**Comment by RodneyBaker** [comment]
Nice one @andeon 
I couldn't resist putting your idea into a scene.

![image](https://github.com/opentoonz/opentoonz/assets/2576432/47da5b5d-2be8-4fbe-8b85-58149e114327)

![pixelgrid](https://github.com/opentoonz/opentoonz/assets/2576432/edb06d19-790d-4fda-9b5c-666e4f353584)

Project file:
[pixelgrid_sampler.zip](https://github.com/opentoonz/opentoonz/files/12375109/pixelgrid_sampler.zip)


Added:  Thanks much for the thought on exporting to spritesheets via sub-cameras... I should have thought of that!

**Comment by andeon** [comment]
Cool @RodneyBaker
the scene with the dog was really funny.

It's not be the ideal solution, but creating a basic template file allows you to create pixel art when needed. @oliver-dreamwalker Not sure what canvas size you use, if it's just for the sprites this gonna work. But for something huge need to create more guides.

I agree that the best approach is a configurable grid for a native function in opentoonz since it to work for many other things.

**Comment by oliver-dreamwalker** [comment]
Yeah, adding a pixel grid template png is a good alternative, I was just curious to know if it was possible implement a grid feature without many issues.
I really saw the potential for pixel animation, with cool effects and all.

**Comment by Mondaley** [comment]
Just add a grid that supports snapping, pixel art, and sprite sheets. It should inherit the features from Godot game engine, PowerPoint, and Aseprite.
Everyone is waiting for this!

Using guidelines to achieve it is cheap. Every program that respects itself should have a genuine customizable GUI for the grid and not an improvised one.
